### DISCUSSION TITLE
History panel should be configurable

### DISCUSSION THREAD

**Discussion by anavarre** [eclipse-theia/theia#11371] [Improvements] [answered]
At **View > History**, you can open up the Git log in the Theia UI. Great feature. However, if you happen to have more than a hundred commits, it becomes extremely slow and makes the IDE interface unusable until all commits and changes within are rendered in the panel.

I could not find a way to make this configurable (in 90% of the cases you really should only care about the 10 latest commits right?) while I'm fairly certain it'd make a huge difference performance wise.

Could it be something for Theia to implement?

**Comment by vince-fugnitto** [comment, marked as answer]
@anavarre thank you for the discussion, from what I understand we should limit `100` entries, and perhaps prompt to load more:

https://github.com/eclipse-theia/theia/blob/0a960a6b200cc6c547a7e403cb21f32fa383001b/packages/scm-extra/src/browser/history/scm-history-contribution.ts#L31

https://github.com/eclipse-theia/theia/blob/0a960a6b200cc6c547a7e403cb21f32fa383001b/packages/scm-extra/src/browser/history/scm-history-widget.tsx#L400-L417

I believe there might be a bug where it is not respected properly at the moment, but we should support the mechanism of passing around the `maxCount`.

**Reply by anavarre** [reply]
Interesting, thanks. Take my 100 threshold with a grain of salt. Perhaps it was exactly 100 commits returned, I didn't count. I just know it took 1-2mn to render for multiple people and the only obvious way to improve the experience is to have a low limit.

What would be ideal is to have ` export const SCM_HISTORY_MAX_COUNT = 10;` and have a text field in the settings where this would be a configurable setting. Alternatively, a dropdown with a few options e.g. 10, 30, 50, 100.

In the Git UI, it would also be a nice to have to be able to click on a "Load more" button or, when you're at the bottom of the panel, scroll down and fetch more commits, similar to an infinite scroll feature if you will.

**Reply by vince-fugnitto** [reply]
> What would be ideal is to have export const SCM_HISTORY_MAX_COUNT = 10; and have a text field in the settings where this would be a configurable setting. Alternatively, a dropdown with a few options e.g. 10, 30, 50, 100.

10 might be a bit low for a default but pull-requests are always welcome!

**Reply by anavarre** [reply]
10 might be low but 100 certainly seems high considering the performance hit. Based on your experience, what would you think would be a good middle ground?

**Reply by vince-fugnitto** [reply]
I've ran the desktop example application locally and did not seem to notice an issue so it might be related to the network.

Note that while I mentioned 100 it does not actually seem to be respected as I highlighted in https://github.com/eclipse-theia/theia/discussions/11371#discussioncomment-3050575 since I can load all commits of the theia repo (back to the initial commit) when the view is first opened.

https://user-images.githubusercontent.com/40359487/176509256-e06624c8-2049-435e-a2fc-7ff267ba763b.mp4

I believe the first part would be to ensure we respect a max, and then we can see if 100 is good or not as a default.

**Reply by anavarre** [reply]
https://github.com/eclipse-theia/theia/issues/11376

**Comment by anavarre** [comment]
Key takeways from this discussion:

- We are hardcoding the Git history max count at 100 but it doesn't seem to be respected ATM. #11376 to address the bug.
- After this bug is fixed, we will be able to revisit whether 100 makes sense or whether we need to aim for a lower hardcoded value.
- Theia maintainers are looking to get external contributions to make this a configurable setting for end users to enter the value that would best suit their needs. https://github.com/eclipse-theia/theia/issues/11378 is the feature request.
